<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>SNP Match</title>
    <script src="https://cdn.tailwindcss.com"></script>
    <style>
        
        #output_table{
            font-family:'Courier New';
        }
        .nucleotide-set-td{
            padding-right:10px;
         
          
        }
        .snp-row{
            height:10px;
            background:#f4f4f4;

        }
        .bold{
            font-weight:bold;
        }
        .text-red-500{
            color:#f56565;
        }
        td * { margin: 0px; }
        table { table-layout: fixed; }
        .snp-row td{
           
            max-width:12px;
        }

        .text-blue-500{
            color:#4299e1;
        }
    </style>
</head>
<body>
   
    <div class="flex items-center flex-col pt-12">
        <div>
            <div class="mt-3 w-64">
                <h3>
                    <div class="font-semibold  ">Search for a gene...</div> <span id="added-files-list"></span>
                </h3>
            </div>
         

            <div class="relative">
                <div class="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none">
                    <svg class="text-gray-500 fill-gray-500" xmlns="http://www.w3.org/2000/svg" height="1em" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license (Commercial License) Copyright 2023 Fonticons, Inc. --><path d="M416 0c17.7 0 32 14.3 32 32c0 59.8-30.3 107.5-69.4 146.6c-28 28-62.5 53.5-97.3 77.4l-2.5 1.7c-11.9 8.1-23.8 16.1-35.5 23.9l0 0 0 0 0 0-1.6 1c-6 4-11.9 7.9-17.8 11.9c-20.9 14-40.8 27.7-59.3 41.5H283.3c-9.8-7.4-20.1-14.7-30.7-22.1l7-4.7 3-2c15.1-10.1 30.9-20.6 46.7-31.6c25 18.1 48.9 37.3 69.4 57.7C417.7 372.5 448 420.2 448 480c0 17.7-14.3 32-32 32s-32-14.3-32-32H64c0 17.7-14.3 32-32 32s-32-14.3-32-32c0-59.8 30.3-107.5 69.4-146.6c28-28 62.5-53.5 97.3-77.4c-34.8-23.9-69.3-49.3-97.3-77.4C30.3 139.5 0 91.8 0 32C0 14.3 14.3 0 32 0S64 14.3 64 32H384c0-17.7 14.3-32 32-32zM338.6 384H109.4c-10.1 10.6-18.6 21.3-25.5 32H364.1c-6.8-10.7-15.3-21.4-25.5-32zM109.4 128H338.6c10.1-10.7 18.6-21.3 25.5-32H83.9c6.8 10.7 15.3 21.3 25.5 32zm55.4 48c18.4 13.8 38.4 27.5 59.3 41.5c20.9-14 40.8-27.7 59.3-41.5H164.7z"/></svg> </div>
                <input type="search"  value="ABCB1" id="geneName" class="mb-12 block w-full p-4 pl-10 text-sm text-gray-900 border border-gray-300 rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 " placeholder="Example: A,B,C" required="">
               
            </div>
        </div>
     
    
    
        <div id="output_table">

        </div>
    </div>
 
</body>

<script src="https://code.jquery.com/jquery-3.7.0.min.js" integrity="sha256-2Pmvv0kuTBOenSvLm6bvfBSSHrUJ+3A7x6P5Ebd07/g=" crossorigin="anonymous"></script>
<script src="https://unpkg.com/@popperjs/core@2"></script>
<script src="https://unpkg.com/tippy.js@6"></script>
<script>

    function getGeneByName(geneName){
        $.ajax({
            url: `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=gene&term=human[orgn],ABCB1[title]&rettype=json&retmode=json`,
            type: "GET",
            dataType: "JSON",
            success: function(response){
                console.log(response);
                const geneId = response.esearchresult.idlist[0];
                console.log('Associated id... ', geneId);
                getGeneSummaryById(geneId,geneName);
            },
            error: function(error){
                console.log(error);
            }
        });
    }

    getGeneByName($("#geneName").val());


    function getGeneSummaryById(geneId,geneName){
        $.ajax({
            url: `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=gene&id=${geneId}&rettype=json&retmode=json`,
            type: "GET",
            dataType: "JSON",
            success: function(response){
              
                const geneObj = response.result[geneId];
                console.log(geneObj);

                const summary = geneObj.summary;
                const description = geneObj.description;
                console.log('Summary... ', description);

                const chrStartPosition = geneObj.chrstart;
                console.log('Start position... ', chrStartPosition);

                const genomicInfo = geneObj.genomicinfo[0];
                const geneStart = genomicInfo.chrstop;
                const geneEnd = genomicInfo.chrstart;

                console.log(chrStartPosition, geneStart, geneEnd, geneEnd-geneStart, chrStartPosition-geneStart);

                const relativeStart = chrStartPosition-geneStart;
                const relativeEnd = geneEnd-chrStartPosition;
                
                //const relativeStartPosition = genomicInfo.chrstart;
                getFASTANucleotideSeqByNamePos(
                    {
                        nuccoreId: genomicInfo.chraccver,
                        geneId: geneId,
                        geneName: geneName,
                        absoluteGeneStart: geneStart
                    },geneStart+1,geneEnd+1);
            },
            error: function(error){
                console.log(error);
            }
        });
    }

    var exceededRowCount = 0;
    function getFASTANucleotideSeqByNamePos(geneInfo,startPosition,endPosition){
        $.ajax({
            url: `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id=${geneInfo.nuccoreId}&rettype=genbank&retmode=text&from=${startPosition}&to=${endPosition}`,
            type: "GET",
            dataType: "text",
            success: function(response){
                const originIndex = response.indexOf("ORIGIN");
                const cleanResponse = response.substring(originIndex+6);
                 const lines = cleanResponse.split("\n");
               
  
                    lines.splice(0,1);
                    console.log(lines[0])
                    console.log(lines[1])
                for (let i = 0; i < lines.length; i++) {
                    const element = lines[i];
  
                    let nucleotides = lines[i].split(" ");
                    //delete all empty nucleotide array items:
                    nucleotides = nucleotides.filter(function (el) {
                        return el != "";
                    });
                    //console.log(nucleotides);
                    const firstIndexOfRow = nucleotides[0];
                    //console.log(firstIndexOfRow);

                   

                    //output to output table, row by row, first item will be firstIndexOfRow, each element will be seperate column
                    let row = `<tr id='nucleotide_row_${firstIndexOfRow}'>`; // Create a new row
                    for(let j = 0; j < nucleotides.length; j++){
                        row += `<td class="nucleotide-set-td">${nucleotides[j]}</td>`; // Add each element as a new column
                    }
                    row += "</tr>"; // Close the row
                   
                    $("#output_table").append(`<tbody id='addn_snp_row_${firstIndexOfRow}'  >${row} </tbody>`); // Add the row to the table
                    
                }

             
                
                getSNPsByGeneName(geneInfo.geneName,geneInfo.geneId, geneInfo.absoluteGeneStart);
                
            },
            error: function(error){
                console.log(error);
            }
        });
    }

    function appendSNPToIndex(index, snp, metadata) {
    
   
    let color = 'blue';
        
    if(index>300 ){
        return false;
    }

    if(index==5){
        console.log(metadata)
    }

    const rowIndex = Math.floor((index - 1) / 60); // when appending, we need to add 1 to mirror DOM
    const relativeColumnIndex = index - (rowIndex * 60);

    
    const relativeCellColumn = Math.floor((relativeColumnIndex - 1) / 10);



    if(metadata.spdi[1]==87503030){
        console.log('index ', index, relativeCellColumn)
    }
    if(metadata.spdi[1]==87503016){
        console.log('index ', index, relativeCellColumn)
    }
   
    // Given that there are 60 nucleotides per row, each separated by a space after every 10

   

    //closest multiple of 60:
    
    const targetRow = Math.floor(index/60)*60+1;
    let targetCellWithinTen = index%10;
    if(targetCellWithinTen == 0){
        targetCellWithinTen = 10;
    }



    // initialize variables to be used in loop.
    var internalRowCountForSNPs = 0; // if SNP in same slot, needs to create a new row
    var hasAppended = false;
    var mulitpleSNPInRow = true; // automatically new line if there is already an SNP in the row

    do {
        var row = $(`#snp_row_${targetRow}_${internalRowCountForSNPs}`); // get first additional SNP row
      
        if(row.length == 0){ // if now additional SNP row, create one with some empty cells
            createSNPRow(targetRow,internalRowCountForSNPs);
            row = $(`#snp_row_${targetRow}_${internalRowCountForSNPs}`);
        }
        
        

        var targetSNPRowColumn = $(row).find('td')[relativeCellColumn+1]; // find all those empty cells
    
        if(targetSNPRowColumn == undefined){ // if there is no empty cell, create a new row
            console.log('SNP Exceeded row count, ', index);
            exceededRowCount++;
            hasAppended = true;
            continue;
        }

        if(snp==''){
            //this means deletion:
            snp = '-';
        }else{
            color = 'red'; // substitution
        }

        if(snp.length>1){
            // multi sub
            
            // get first char of snp
            if(snp.length==metadata.spdi[2].length){
                console.log(snp.length, metadata.spdi[2].length)
                snp = '=';
            }
            else if(snp.length<metadata.spdi[2].length){
                snp = '-'
            }
            else{
                snp = '+';
            }
            

        }

        let tippyObj = {
                content: `${metadata.spdi[0]}, ${metadata.spdi[1]} (${index})<br/> 
                ${metadata.spdi[2]}
                <br/>
                ${metadata.spdi[3]}
                <br/>
                rs${metadata.uid} <a href="https://www.ncbi.nlm.nih.gov/snp/rs${metadata.uid}" class="underline" target="_blank">View on NCBI</a>`,
                placement: 'bottom',
                allowHTML: true,
                interactive: true,
                };

        if($(targetSNPRowColumn).html() == ''){ // if it's a brand new row, just find the correct slot, and and yyyyXyyy, X in the correct row
            for(l=1; l<=10; l++){
                if(l==targetCellWithinTen){
                    if(snp=='-' || snp=='+'){
                        color = 'blue';
                    }
                    $(targetSNPRowColumn).append(`<span class="snp bold text-${color}-500"  id="snp_${index}_${internalRowCountForSNPs}">${snp}</span>`);

             
                tippy(`#snp_${index}_${internalRowCountForSNPs}`, tippyObj);


                    hasAppended = true;
                }else{
                    $(targetSNPRowColumn).append(`<span>.</span>`);
                }
            }
        }
        else{ // there is an SNP in this row, find out if it's in the target slot or not..
            for(l=1; l<=10; l++){
                if(l==targetCellWithinTen){
                    let targetCellSpan = targetSNPRowColumn.querySelectorAll('span')[l-1];
                    
                    if($(targetCellSpan).html()=='.' && mulitpleSNPInRow==true){ // NOT in target slot, we are good to go, now there are multiple SNPs in the same row
                        if(snp=='-' || snp=='+'){
                            color = 'blue';
                        }
                        $(targetCellSpan)[0].outerHTML = `<span class="snp bold text-${color}-500"  id="snp_${index}_${internalRowCountForSNPs}">${snp}</span>`;

                        tippy(`#snp_${index}_${internalRowCountForSNPs}`, tippyObj);


                        hasAppended = true;
                    }else{ // target slot! Skip to next row
                        internalRowCountForSNPs++;
                    }
                }
            }
        }
    } while (hasAppended==false);   
  

    
    
}

function createSNPRow(targetRow,internalRowCountForSNPs){
    
    $(`#addn_snp_row_${targetRow}`).append(`<tr id='snp_row_${targetRow}_${internalRowCountForSNPs}' class=" snp-row" data-id="${targetRow}_${internalRowCountForSNPs}">
                        <td ></td>
                        <td></td>
                        <td></td>
                        <td></td>
                        <td></td>
                        <td></td>    
                        <td></td>  
                    </tr>`); // Add the row to the table
    
        //refresh DOM to reflect changes
}

function getSNPsByGeneName(geneName,geneId,geneStart){
  

    $.ajax({
            url: `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=snp&term=${geneId},${geneName},Homo+sapiens,3_prime_UTR_variant&retmax=500&rettype=json&retmode=json`,
            type: "GET",
            dataType: "json",
            success: function(response){
                console.log(response);
                let idList = response.esearchresult.idlist.join(',');
                // keep first 3
              
                
                console.log(idList)
                fetchSNPData(idList, geneStart)
            }
        
        });

    
}

function fetchSNPData(snpIdArray,geneStart){
   
    $.ajax({
            url: `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=snp&rettype=json&retmode=json`,
            type: "POST",
            data: {
                id: snpIdArray
            },
            dataType: "json",
            success: function(response){
                console.log(response);
                // loop through result object key-value for in:
                const snps = response.result;

                for (const snpId in snps){
                    snp = snps[snpId];
                  
                    // break it up by :

                    //split by , 
                    if(snp.spdi == undefined){
                        continue;
                    }
                    const snp_parent_array = snp.spdi.split(',');

                    for(let i = 0; i < snp_parent_array.length; i++){
                        const spdi = snp_parent_array[i].split(':');
                    

                        const chromosomeId = spdi[0];
                        const position = spdi[1];
                        const ref = spdi[2];
                        const alt = spdi[3];
                        //check if alt contains string NC_000007.14
                      
                        const relativePosition = position-geneStart+1;
       

                        // if(ref.length>1 || alt.length>1){
                        //     console.log('ref is more than 1');
                        //     continue;
                        // }

                        const metadata = {
                            spdi: spdi,
                            uid: snp.uid,
                        }

                        appendSNPToIndex(relativePosition, alt, metadata);
                
                    }
                }
                

                
            }
        
        });   
}





</script>
</html>