<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>SNP Match</title>
    <script src="https://cdn.tailwindcss.com"></script>
    <style>
        
        #output_table{
            font-family:'Courier New';
        }
        .nucleotide-set-td{
            padding-right:10px;
         
          
        }
        .snp-row{
            height:10px;
            background:#f4f4f4;

        }

        .snp-line-row{
            
        }
        .del-row{
            height:10px;
          
           
        }
        .dup-row{
            height:10px;
         
        }
        .bold{
            font-weight:bold;
        }
        .text-red-500{
            color:#f56565;
        }
        td * { margin: 0px; }
        table { table-layout: fixed; }
        .snp-row td{
           
            max-width:12px;
        }

        .text-blue-500{
            color:#4299e1;
        }

        .snp{
            width:1ch;
            height: 1ch;
            display:inline-flex;
        }
        .spacer{
            display:invisible;
            color:#f4f4f4;
        }
        .line-td{
           
  
        }

       
    </style>
</head>
<body>
    <div class="fixed bg-black h-screen w-full z-30 bg-opacity-75 hidden" id="loader">
        <div role="status" class="absolute -translate-x-1/2 -translate-y-1/2 top-2/4 left-1/2">
            <svg aria-hidden="true" class="w-8 h-8 mr-2 text-gray-200 animate-spin dark:text-gray-600 fill-blue-600" viewBox="0 0 100 101" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z" fill="currentColor"/><path d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z" fill="currentFill"/></svg>
            <span class="sr-only">Loading...</span>
        </div>
    </div>
    <div class="max-w-5xl mx-auto py-12">
        <h2 class="text-4xl font-extrabold ">SNP mRNA Match Tool</h2>
        <p class="my-4 text-lg text-gray-500">A simple tool to remove selected columns from a series of .xlsx files.
            Made by Martin Bouldo.</p>
   
    <div class="flex  flex-col ">
        <div>
            <div class="flex space-x-6">
                <div>
                    <div class="mt-3 w-64">
                        <h3>
                            <div class="font-semibold  mb-1">Search for a gene...</div> 
                        </h3>
                    </div>
                    <div class="relative">
                        <div class="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none">
                            <svg class="text-gray-500 fill-gray-500" xmlns="http://www.w3.org/2000/svg" height="1em" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license (Commercial License) Copyright 2023 Fonticons, Inc. --><path d="M416 0c17.7 0 32 14.3 32 32c0 59.8-30.3 107.5-69.4 146.6c-28 28-62.5 53.5-97.3 77.4l-2.5 1.7c-11.9 8.1-23.8 16.1-35.5 23.9l0 0 0 0 0 0-1.6 1c-6 4-11.9 7.9-17.8 11.9c-20.9 14-40.8 27.7-59.3 41.5H283.3c-9.8-7.4-20.1-14.7-30.7-22.1l7-4.7 3-2c15.1-10.1 30.9-20.6 46.7-31.6c25 18.1 48.9 37.3 69.4 57.7C417.7 372.5 448 420.2 448 480c0 17.7-14.3 32-32 32s-32-14.3-32-32H64c0 17.7-14.3 32-32 32s-32-14.3-32-32c0-59.8 30.3-107.5 69.4-146.6c28-28 62.5-53.5 97.3-77.4c-34.8-23.9-69.3-49.3-97.3-77.4C30.3 139.5 0 91.8 0 32C0 14.3 14.3 0 32 0S64 14.3 64 32H384c0-17.7 14.3-32 32-32zM338.6 384H109.4c-10.1 10.6-18.6 21.3-25.5 32H364.1c-6.8-10.7-15.3-21.4-25.5-32zM109.4 128H338.6c10.1-10.7 18.6-21.3 25.5-32H83.9c6.8 10.7 15.3 21.3 25.5 32zm55.4 48c18.4 13.8 38.4 27.5 59.3 41.5c20.9-14 40.8-27.7 59.3-41.5H164.7z"/></svg> </div>
                        <input type="search"  value="MET" id="geneName" class=" block w-full p-2.5 pl-10 text-sm text-gray-900 border border-gray-300 rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 " placeholder="Example: ABCB1" required="">
                    
                    </div>
                </div>
                <div class="flex items-end">
                    <div class=" mb-1">
                        <button onclick="getGeneByName($('#geneName').val());" class="px-4 py-2 
                        text-sm font-medium tracking-wide text-white 
                        capitalize bg-blue-500 bg-blue-700 hover:bg-blue-800 focus:ring-4 
                        focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-4 py-2">
                            Search</button>
                    </div>
                </div>
                <div>
                  

                </div>

                
            </div>
            <div class="flex">
                <div >
                    <div class="mt-3 w-128">
                        <h3>
                            <div class="font-semibold  mb-1">Output the following mRNA Transcript Version...</div> <span id=""></span>
                        </h3>
                    </div>
                    <select id="transcriptNames" class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-96 p-2.5 pt-3 ">
                    </select>
                </div>
                <div>
                    <div class="flex items-center mt-12 ml-12">
                        <input  id="isReversed" type="checkbox" value="" class="w-4 h-4 text-blue-600 bg-gray-100 border-gray-300 rounded focus:ring-blue-500 dark:focus:ring-blue-600 dark:ring-offset-gray-800 focus:ring-2 dark:bg-gray-700 dark:border-gray-600">
                        <label for="isReversed" class="ml-2 text-sm font-medium text-gray-900 " id="">Reversed</label>
                    </div>
                </div>
            </div>
            <div>
                <div class="mt-3 ">
                    <h3>
                        <div class="font-semibold  mb-1">Transcript Metadata:</div> 
                        <div class="mb-3">
                            <div class="block">CDS (Starting Point): <span id="cds_output"></span></div>
                        </div>
                        <hr/>
                        <div  class="text-mono ]">
                            <pre id="original_response" style="height:300px;" class="overflow-y-scroll"></pre>

                            <div class="flex justify-end items-center bg-white mt-[-20px] bg-opacity-75"><div id="expandResponse" class="text-brand transition flex items-center text-blue-500 cursor-pointer"><svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="2" stroke="currentColor" aria-hidden="true" class="h-5 w-5 mr-1"><path stroke-linecap="round" stroke-linejoin="round" d="M4 8V4m0 0h4M4 4l5 5m11-1V4m0 0h-4m4 0l-5 5M4 16v4m0 0h4m-4 0l5-5m11 5l-5-5m5 5v-4m0 4h-4"></path></svg> Expand</div></div>

                            <div class="flex justify-end items-center bg-white mt-[-20px] bg-opacity-75"><div id="collapseResponse" class="hidden text-brand transition flex items-center text-blue-500 cursor-pointer"><svg class="h-5 w-5 mr-1" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" class="w-6 h-6">
                                <path stroke-linecap="round" stroke-linejoin="round" d="M9 9V4.5M9 9H4.5M9 9L3.75 3.75M9 15v4.5M9 15H4.5M9 15l-5.25 5.25M15 9h4.5M15 9V4.5M15 9l5.25-5.25M15 15h4.5M15 15v4.5m0-4.5l5.25 5.25" />
                              </svg>
                               Collapse</div></div>
                        </div>
                    </h3>
                </div>
            </div>
           
        </div>
     
     
        <div>
            <div class="">
                <h2 class="text-2xl font-extrabold mt-12">mRNA alignments</h2>
               
            </div>
           
<textarea name="mRNA Alignemnt"  
value=""

rows="3" class="w-full border border-gray-400" id="mRNA_alignment"></textarea>
        </div>

        <div class="mt-3 flex h-10 justify-between">
            <div class="flex">
                <a href="https://franziskaahrend.pythonanywhere.com/complementaryAlignment" target="_blank" class="px-4
                text-sm font-medium tracking-wide text-white
                capitalize bg-blue-500 bg-blue-700 hover:bg-blue-800 focus:ring-4
                focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-4 flex items-center">
                        Go To Ahrend Complementary alignment Tool
                    </a>
                    <a href="#!" onclick="copyFasta()" class="px-4   ml-3
                text-sm font-medium tracking-wide text-white
                capitalize bg-blue-500 bg-blue-700 hover:bg-blue-800 focus:ring-4
                focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-4 flex items-center">
                        Copy Fasta Sequence
                    </a>
            </div>

    <a href="#!" onClick="prependmRNAToIndex()" class="px-4   ml-3 
            text-sm font-medium tracking-wide text-white 
            capitalize bg-blue-500 bg-green-700 hover:bg-blue-800 focus:ring-4 
            focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-4 flex items-center">
        Add Alignment
        </a>
    
        </div>

        <div>
            <h2 class="text-2xl font-extrabold mt-12">SNP Viewer</h2>
        </div>
        <div id="output_table" class="mt-5">

        </div>
        <input type="text" value="" id="copyInput">
    </div>
</div>
</body>

<script src="https://code.jquery.com/jquery-3.7.0.min.js" integrity="sha256-2Pmvv0kuTBOenSvLm6bvfBSSHrUJ+3A7x6P5Ebd07/g=" crossorigin="anonymous"></script>
<script src="https://unpkg.com/@popperjs/core@2"></script>
<script src="https://unpkg.com/tippy.js@6"></script>
<script>

    var orderMode = 'normal';


    var totalLength = 0;

    function showLoader(){
        $("#loader").show();
    }
    function hideLoader(){
        $("#loader").hide();
    }

    function getGeneByName(geneName){
        showLoader();
        $.ajax({
            url: `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=gene&term=human[orgn],${geneName}[title]&rettype=json&retmode=json`,
            type: "GET",
            dataType: "JSON",
            success: function(response){
                console.log(response);
                const geneId = response.esearchresult.idlist[0];
                console.log('Associated id... ', geneId);
                getGeneSummaryById(geneId,geneName);
              
            },
            error: function(error){
                console.log(error);
            }
        });
    }

   


    function getGeneSummaryById(geneId,geneName){
        $.ajax({
            url: `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=gene&id=${geneId}&rettype=json&retmode=json`,
            type: "GET",
            dataType: "JSON",
            success: function(response){
              
                const geneObj = response.result[geneId];
                console.log(geneObj);

                const summary = geneObj.summary;
                const description = geneObj.description;

                const chrStartPosition = geneObj.chrstart;

                const genomicInfo = geneObj.genomicinfo[0];
                const geneStart = genomicInfo.chrstop;
                const geneEnd = genomicInfo.chrstart;

                console.log(chrStartPosition, geneStart, geneEnd, geneEnd-geneStart, chrStartPosition-geneStart);

                const relativeStart = chrStartPosition-geneStart;
                const relativeEnd = geneEnd-chrStartPosition;
                
                //const relativeStartPosition = genomicInfo.chrstart;
                getFASTANucleotideSeqByNamePos(
                    {
                        nuccoreId: genomicInfo.chraccver,
                        geneId: geneId,
                        geneName: geneName,
                        absoluteGeneStart: geneStart
                    },geneStart+1,geneEnd+1);
            },
            error: function(error){
                console.log(error);
            }
        });
    }

    var exceededRowCount = 0;
    function getFASTANucleotideSeqByNamePos(geneInfo,startPosition,endPosition){
        $.ajax({
            url: `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id=${geneInfo.nuccoreId}&rettype=genbank&retmode=text&from=${startPosition}&to=${endPosition}`,
            type: "GET",
            dataType: "text",
            success: function(response){
                const transcript_ids = extractTranscriptIds(response);
                const originIndex = response.indexOf("ORIGIN");
                const cleanResponse = response.substring(originIndex+6);
         
              
                const lines = cleanResponse.split("\n");



               
             
                
                getSNPsByGeneName(geneInfo.geneName,geneInfo.geneId, geneInfo.absoluteGeneStart);
                getmRNATranscriptsByGeneId(geneInfo.geneId);
                fetchSelectedTranscript();
                
            },
            error: function(error){
                console.log(error);
            }
        });
    }

    var gene_transcript_names = [];
    function getmRNATranscriptsByGeneId(geneId){
        $.ajax({
            url: `https://api.ncbi.nlm.nih.gov/datasets/v2alpha/gene`,
            type: "POST",
            async: false,
            data:JSON.stringify({
                    "geneIds": [
                        geneId
                    ],
                    "returned_content": "PRODUCT"
                }), 
                contentType: "application/json; charset=utf-8",
            dataType: "JSON",
            success: function(response){

                const transcripts = response.reports[0].product.transcripts;
                gene_transcript_names = []
                for(transcript of transcripts){
                    gene_transcript_names.push(transcript.accession_version)
                    $("#transcriptNames").append(`<option>${transcript.accession_version}</option>`)
                }
         
          
            },
            error: function(error){
                console.log(error);
            }
        });
    }
    var offset;

    function getFinalIndexPoint(text){
     
        const regex = /(\d+) bp/;
        const match = text.match(regex);

        if (match && match[1]) {
        const basePairs = parseInt(match[1]);
        console.log(`Extracted ${basePairs} bp from the text.`);
        return basePairs
        } else {
        console.log("No bp value found in the text.");
        }
    }

    var currentTranscript = {
        fasta: '',
        geneBank: ''
    };

    function fetchSelectedTranscript(){
        showLoader();
        $("#output_table").empty();
        addedItems = [];
        $("#output_table").html('');
        const transcript_name = $("#transcriptNames").val();
        $.ajax({
            url: `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id=${transcript_name}&rettype=genbank&retmode=text`,
            type: "GET",
            dataType: "text",
            success: function(response){
                $("#original_response").text(response);
                const allLines = response.split("\n");
                reversedOffsetIndex = getFinalIndexPoint(allLines[0]);
                const transcript_ids = extractTranscriptIds(response);
                
                const originIndex = response.indexOf("ORIGIN");
                const cleanResponse = response.substring(originIndex+6);
                var lines = cleanResponse.split("\n");
               

                const cdsIndex = response.indexOf("CDS");
                const cleanCDS = response.substring(cdsIndex+6);
                $("#cds_output").text(cleanCDS.split("\n")[0]);



                 offset = cleanCDS.split("\n")[0].split('..')[1];

               
               
                lines.splice(0,1);

                if(orderMode == 'reversed'){
                    console.log(lines)
                    lines = reverseSequence(lines);
                    console.log(lines)
                }

                let fasta = lines;
                // remove numbers, spaces, and from array and join back together
                fasta = fasta.map(function (el) {
                    return el.replace(/[^atcg]/ig, '');
                });

                currentTranscript.fasta = fasta.join('');
                

   
                for (let i = 0; i < lines.length; i++) {
                    const element = lines[i];
  
                    let nucleotides = lines[i].split(" ");
                    //delete all empty nucleotide array items:
                    nucleotides = nucleotides.filter(function (el) {
                        return el != "";
                    });
                    const firstIndexOfRow = nucleotides[0];

                    //output to output table, row by row, first item will be firstIndexOfRow, each element will be seperate column
                    let row = `<tr id='nucleotide_row_${firstIndexOfRow}'>`; // Create a new row
                    for(let j = 0; j < nucleotides.length; j++){
                        row += `<td class="nucleotide-set-td">${nucleotides[j]}</td>`; // Add each element as a new column
                    }
                    row += "</tr>"; // Close the row
                   
                    $("#output_table").append(`<tbody table cellspacing="1" cellpadding="1" id='addn_snp_row_${firstIndexOfRow}'  >${row} </tbody>
                    <tbody table cellspacing="1" cellpadding="1" id='addn_del_row_${firstIndexOfRow}'  ></tbody>
                    <tbody table cellspacing="1" cellpadding="1" id='addn_dup_row_${firstIndexOfRow}'  ></tbody>`); // Add the row to the table
                }

                markupStoredSNPData();
                hideLoader();

            },
            error: function(error){
                console.log(error);
            }
        });
    }

    function reverseSequence(obj) {
    let sequence = '';

    // Extract sequences and concatenate
    for (let key in obj) {
        sequence += obj[key].replace(/[^atcg]/ig, '');
    }

    // Reverse the sequence
    let reversedSequence = sequence.split('').reverse().join('');
    console.log(reversedSequence)
    totalLength = reversedSequence.length;

    // Split the reversed sequence into chunks of 60 for the output
    let chunks = [];
    for (let i = 0; i < reversedSequence.length; i += 60) {
        let chunk = reversedSequence.substr(i, 60);
        
        // Add spaces every 10 characters
        let spacedChunk = '';
        for (let j = 0; j < chunk.length; j += 10) {
            spacedChunk += chunk.substr(j, 10) + ' ';
        }
        chunks.push(spacedChunk.trim());
    }

    // Convert chunks into the desired format
    let result = {};
    for (let i = 0; i < chunks.length; i++) {
        let startNumber = (i * 60) + 1;
        result[i] = `${("        " + startNumber).slice(-8)} ${chunks[i]}`;
    }

//    return result;
//return result as standard array

    return Object.values(result);

}

    $("#transcriptNames").change(function(){
        if($("#transcriptNames").val() == ""){
            return;
        }
        
        
        fetchSelectedTranscript();
    })

    //check if #isReversed is checked
  
    $(document).ready(function() {
        $('#isReversed').change(function() {
            if(this.checked) {
                orderMode = 'reversed';
            }else{
                orderMode = 'normal';
            }
            fetchSelectedTranscript();    
        });
    });


    function extractTranscriptIds(genbankText) {
    // Create a regex pattern to find mRNA entries and transcript_ids
        const regex = /mRNA[\s\S]*?\/transcript_id="([^"]*)"/g;
        let match;
        let transcriptIds = [];

        // Find matches and push transcript ids to the array
        while ((match = regex.exec(genbankText)) !== null) {
            transcriptIds.push(match[1]);
        }

        // Return the array of transcript ids
        return transcriptIds;
    }


    $("#expandResponse").click(function(){
        $("#original_response").css('height', 'auto');
        $("#expandResponse").hide();
        $("#collapseResponse").show();
    })
    $("#collapseResponse").click(function(){
        $("#original_response").css('height', '300px');
        $("#expandResponse").show();
        $("#collapseResponse").hide();
    })
    

    function prependmRNAToIndex(alignment){
        const data = getAlignmentValue(alignment);
        const mRNA_seq = data.mRNA_line.replace(/\s/g, '');
        const mRNA_seq_length = mRNA_seq.length;  

        let lineSeq = data.matchLine;
        console.log(lineSeq);
        // delete first n numbers of chars (data.mRNA_leadingWhiteSpace)
        lineSeq = lineSeq.substring(data.mRNA_leadingWhiteSpace);
        console.log(lineSeq)

        console.log('seq len', mRNA_seq_length)

        let spacerSet = false;
        let j = 0;
        for(let i=0; i<data.seq_length; i++){
            j++;
            const index = Number(data.mRNA_leadingWhiteSpace)+j;
            

          

            const rowIndex = Math.floor((index - 1) / 60); // when appending, we need to add 1 to mirror DOM
            const relativeColumnIndex = index - (rowIndex * 60);
            const relativeCellColumn = Math.floor((relativeColumnIndex - 1) / 10);
            let targetRow = Math.floor(index/60)*60+1;
            if(targetRow==(index/60)*60+1){
                targetRow = Math.floor((index-1)/60)*60+1
            }
            let targetCellWithinTen = index%10;
            if(index%10==0 && index!=0){
                targetCellWithinTen = 10;
            }else{
               
            }

           

            console.log('index: ', i, index, targetRow, relativeCellColumn+1, targetCellWithinTen, lineSeq.charAt(i))

            var row = $(`#mRNA_row_${targetRow}_0`); // get first additional SNP row
        
            if(row.length == 0){ // if now additional SNP row, create one with some empty cells
                prependLineRow(targetRow,0,'line');
                prependmRNARow(targetRow,0,'mRNA');
                row = $(`#mRNA_row_${targetRow}_0`);
            }
            
            var targetSNPRowColumn = $(row).find('td')[relativeCellColumn+1]; // find all those empty cells
          
            //console.log(targetRow, relativeCellColumn+1, targetCellWithinTen)
            $(`#mrna_tdspan_${targetRow}_${relativeCellColumn+1}_${targetCellWithinTen}`).html(`${mRNA_seq.charAt(i)}`);
            $(`#line_tdspan_${targetRow}_${relativeCellColumn+1}_${targetCellWithinTen}`).html(`${lineSeq.charAt(i) ==' ' ? '.' : lineSeq.charAt(i)}`);

            // remove select-none spacer classes from both tdSpans
            $(`#mrna_tdspan_${targetRow}_${relativeCellColumn+1}_${targetCellWithinTen}`).removeClass('select-none');
            $(`#mrna_tdspan_${targetRow}_${relativeCellColumn+1}_${targetCellWithinTen}`).removeClass('spacer');

            if(lineSeq.charAt(i) !=' '){
                $(`#line_tdspan_${targetRow}_${relativeCellColumn+1}_${targetCellWithinTen}`).removeClass('select-none');
                $(`#line_tdspan_${targetRow}_${relativeCellColumn+1}_${targetCellWithinTen}`).removeClass('spacer');
            }

            
        }

    }


    function appendSNPToIndex(index, snp, snpObject) {

    const delta = snpObject.end-snpObject.start;

    let outerWidth = snpObject.end-snpObject.start;
    if(outerWidth==0){
        outerWidth = 1;
    }else{
        outerWidth+=1;
    }

        if(orderMode=='reversed'){
          
            index = reversedOffsetIndex - index + 1;
            if(outerWidth>1){
                console.log(index, outerWidth)
                index = index - outerWidth+1;
                console.log(index, outerWidth)
            }
        }
        else{
            if(outerWidth>1){
                // index = index-1;
            }
        }
    let color = 'blue';

    const rowIndex = Math.floor((index - 1) / 60); // when appending, we need to add 1 to mirror DOM
    const relativeColumnIndex = index - (rowIndex * 60);
    const relativeCellColumn = Math.floor((relativeColumnIndex - 1) / 10);
    const targetRow = Math.floor(index/60)*60+1;
    let targetCellWithinTen = index%10;

    if(targetCellWithinTen == 0){
        targetCellWithinTen = 10;
    }

   // console.log(index, rowIndex, relativeCellColumn, targetRow, targetCellWithinTen, snp, snpObject)

    var snpType = snpObject.snpType ;
    if(snpType==null){
        snpType = 'del';
        snpObject.snpType = 'del';
        snpObject.isDel = true;
      
    }

   

    // initialize variables to be used in loop.
    var internalRowCountForSNPs = 0; // if SNP in same slot, needs to create a new row
    var hasAppended = false;
    var mulitpleSNPInRow = false; // automatically new line if there is already an SNP in the row

    do {
        var row = $(`#${snpObject.snpType}_row_${targetRow}_${internalRowCountForSNPs}`); // get first additional SNP row
      
        if(row.length == 0){ // if now additional SNP row, create one with some empty cells
            createAddnRow(targetRow,internalRowCountForSNPs,snpObject.snpType);
            row = $(`#${snpObject.snpType}_row_${targetRow}_${internalRowCountForSNPs}`);
        }
        
        
        var targetSNPRowColumn = $(row).find('td')[relativeCellColumn+1]; // find all those empty cells
    
        if(targetSNPRowColumn == undefined){ // if there is no empty cell, create a new row
            exceededRowCount++;
            hasAppended = true;
            continue;
        }


      
        

        
        if(outerWidth+targetCellWithinTen>11){
            outerWidth+=1;
            if(outerWidth+targetCellWithinTen>21){
                outerWidth+=1;
            }
        }

        if(orderMode=='reversed' && outerWidth>=1){
            index =  index-outerWidth;
        }

        const typeTextMap = {
            snp: 'Substitution',
            del: 'Deletion',
            dup: 'Duplication',
            ins: 'Insertion',
        }

        let tippyObj = {
                content: `(${index}) - ${delta+1}, ${typeTextMap[snpType]}<br/> 
                ${snpObject.string}<br/>
                rs${snpObject.uid} <a href="https://www.ncbi.nlm.nih.gov/snp/rs${snpObject.uid}" class="underline" target="_blank">View on NCBI</a>`,
                placement: 'bottom',
                allowHTML: true,
                interactive: true,
        };

        const typeColor = {
            snp: 'red',
            del: 'blue',
            dup: 'blue',
            ins: 'blue',
        }
       

        let additionalClass = '';
        if(snpType=='del' || snpType=='dup' || snpType=='ins'){
            additionalClass = `absolute`;
        }


        // bg-blue-500, bg-red-500, bg-green-500

        if($(targetSNPRowColumn).html() == ''){ // if it's a brand new row, just find the correct slot, and and yyyyXyyy, X in the correct row
            for(l=1; l<=10; l++){
                if(l==targetCellWithinTen){
                    if(snp=='-' || snp=='+'){
                        color = 'blue';
                    }
                    $(targetSNPRowColumn).append(`<span class="snp bold text-${color}-500 rail-box-${snpType} items-center justify-center ${additionalClass}"  id="${snpType}_${index}_${internalRowCountForSNPs}" data-id="${snpObject.uid}" style="width:${outerWidth}ch;">
                        <span class="bg-${typeColor[snpType]}-500 h-1 w-full"></span>
                        </span>`);

             
                
               tippy(`#${snpType}_${index}_${internalRowCountForSNPs}`, tippyObj);

                    hasAppended = true;
                }else{
                    $(targetSNPRowColumn).append(`<span class='select-none spacer'>.</span>`);
                }
            }
        }
        else{ // there is an SNP in this row, find out if it's in the target slot or not..
            for(l=1; l<=10; l++){
                if(l==targetCellWithinTen){
                    let targetCellSpan = targetSNPRowColumn.querySelectorAll('span')[l-1];
                    
                    if($(targetCellSpan).html()=='.' && mulitpleSNPInRow==true){ // NOT in target slot, we are good to go, now there are multiple SNPs in the same row
                        if(snp=='-' || snp=='+'){
                            color = 'blue';
                        }
                        $(targetCellSpan)[0].outerHTML = `<span class="snp bold text-${color}-500 rail-box-${snpType} ${additionalClass}"  items-center justify-center"  id="${snpType}_${index}_${internalRowCountForSNPs}" data-id="${snpObject.uid}" style="width:${outerWidth}ch;">
                        <span class="bg-${typeColor[snpType]}-500 h-1 w-full"></span>
                        </span>`;

                        tippy(`#${snpType}_${index}_${internalRowCountForSNPs}`, tippyObj);


                        hasAppended = true;
                    }else{ // target slot! Skip to next row
                        if(snpObject.snpType=='snp'){
                            hasAppended = true;
                            continue;
                        }
                        internalRowCountForSNPs++;
                    }
                }
            }
        }
    } while (hasAppended==false);   
  

    
    
}

function createAddnRow(targetRow, subindex, type){
    
    $(`#addn_${type}_row_${targetRow}`).append(`<tr id='${type}_row_${targetRow}_${subindex}' class=" snp-row ${type}-row" data-id="${targetRow}_${subindex}">
                        <td ></td>
                        <td></td>
                        <td></td>
                        <td></td>
                        <td></td>
                        <td></td>    
                        <td></td>  
                    </tr>`); // Add the row to the table
    
}




function prependmRNARow(targetRow, subindex, type){
    $(`#addn_snp_row_${targetRow}`).prepend(`<tr id='${type}_row_${targetRow}_${subindex}' class=" snp-row ${type}-row mt-4" data-id="${targetRow}_${subindex}">
                        <td>${tenSpans(targetRow, 0, 'mrna')}</td>
                        <td>${tenSpans(targetRow, 1, 'mrna')}</td>
                        <td>${tenSpans(targetRow, 2, 'mrna')}</td>
                        <td>${tenSpans(targetRow, 3, 'mrna')}</td>
                        <td>${tenSpans(targetRow, 4, 'mrna')}</td>
                        <td>${tenSpans(targetRow, 5, 'mrna')}</td>    
                        <td>${tenSpans(targetRow, 6, 'mrna')}</td>  
                    </tr>`); // Add the row to the table

}

function prependLineRow(targetRow, subindex, type){
    $(`#addn_snp_row_${targetRow}`).prepend(`<tr id='${type}_row_${targetRow}_${subindex}' class=" snp-row snp-line-row ${type}-row mt-[-5px]" data-id="${targetRow}_${subindex}">
                        <td class="line-td">${tenSpans(targetRow, 0, 'line')}</td>
                        <td class="line-td">${tenSpans(targetRow, 1, 'line')}</td>
                        <td class="line-td">${tenSpans(targetRow, 2, 'line')}</td>
                        <td class="line-td">${tenSpans(targetRow, 3, 'line')}</td>
                        <td class="line-td">${tenSpans(targetRow, 4, 'line')}</td>
                        <td class="line-td">${tenSpans(targetRow, 5, 'line')}</td>    
                        <td class="line-td">${tenSpans(targetRow, 6, 'line')}</td>  
                    </tr>`); // Add the row to the table

}

function tenSpans(targetRow,index, type){
    let spans = '';
    let max = 10;
    if(index==0){
        max = 5;
    }
    for(let i=1; i<=max; i++){
        spans += `<span class="select-none spacer" id="${type}_tdspan_${targetRow}_${index}_${i}">.</span>`;
    }
    return spans;
}

function getSNPsByGeneName(geneName,geneId,geneStart){
  

    $.ajax({
            url: `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=snp&term=${geneId},${geneName},Homo+sapiens,3_prime_UTR_variant&retmax=500&rettype=json&retmode=json`,
            type: "POST",
            dataType: "json",
            async: false,
            success: function(response){
                let idList = response.esearchresult.idlist.join(',');
                fetchSNPData(idList, geneStart)
            }
        
        });

    
}

var addedItems = [];

var snpData;

function fetchSNPData(snpIdArray){
   
    $.ajax({
            url: `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=snp&rettype=json&retmode=json`,
            type: "POST",
            async: false,
            data: {
                id: snpIdArray
            },
            dataType: "json",
            success: function(response){
                snpData = response.result; // called in markupStoredSNPData when needed (transcript fetch complete...)
            }
        });   
}
var count = 0
function markupStoredSNPData(){
    const snps = snpData;

    for (const snpId in snps){
        snp = snps[snpId];
    
        if(snp.docsum == undefined){
            continue;
        }
        const docsum = snp.docsum.split(',');
        //console.log(docsum)
        for(varsnp of docsum){
            count++;
            // if contains HGVS continue;
            if(varsnp.includes('HGVS')){
                continue;
            }
            // break it up by :
            splitVarSnp = varsnp.split(':');
            if(splitVarSnp[0]==gene_transcript_names[3]){
                const snpObject = handleSNPTextData(splitVarSnp[1]);
            
                const spdi = snp.spdi.split(',')[0].split[':'];
                if(snp.spdi == undefined){
                    continue;
                }   

                snpObject.spdi = snp.spdi;
                snpObject.uid = snp.uid;
                
                if(snpObject.original==null){
                    snpObject.original = 'x';
                }

                if(snpObject.isDel || snpObject.isDup || snpObject.isIns){
                    if(addedItems.includes(`${snpObject.start}_${snpObject.end}_${snpObject.snpType}`)){
                        continue;
                    }else{
                        addedItems.push(`${snpObject.start}_${snpObject.end}_${snpObject.snpType}`);
                    }
                }
                
                appendSNPToIndex(parseInt(snpObject.start)+parseInt(offset), snpObject.original, snpObject);
            }

        }
    }
}

function handleSNPTextData(snpString) {
    if(snpString==undefined){
        return;
    }
    // Initialize object with default values
    let snpStringObject = {
        start: null,
        end: null,
        isSub: false,
        isDel: false,
        isDup: false,
        isIns: false,
        original: null,
        snpType: null,
        string: snpString,
    };

    // Extract start and end position using regular expressions

    const position = snpString.match(/\d+(_\d+)?/)[0].split('_');

    //check if snpstring even contains *
    let original;
    let parts;
    if(snpString.includes('*')==false){
       original = snpString.split('>')[0].match(/[A-Z]+/);
       parts = snpString.split('.');
    }else{
         original = snpString.split('>')[0].split('*')[1].match(/[A-Z]+/);
          parts = snpString.split("*");
    }

    
    if(original){
        snpStringObject.original = original[0];
    }

    if(original=='SEQ'){
        original==null;
    }
    



 

// Get the start and end positions of the snpString.
    const start = parts[1].match(/\d+(_\d+)?/)[0]
    end = start;
    if(parts[2]){
        end = parts[2].match(/\d+(_\d+)?/)[0]
    }

    snpStringObject.start = start
    snpStringObject.end = end


    // Check snpString type
    if (snpString.includes('>')) {
        snpStringObject.isSub = true;
        snpStringObject.snpType = 'snp';
    } else if (snpString.includes('del')) {
        snpStringObject.isDel = true;
        snpStringObject.snpType = 'del';
    } else if (snpString.includes('dup')) {
        snpStringObject.isDup = true;
        snpStringObject.snpType = 'dup';
    } else if (snpString.includes('ins')) {
        snpStringObject.isIns = true;
        snpStringObject.snpType = 'ins';
    }

    if(snpStringObject.snpType=='ins'){
        console.log('ins');
    }
    

    return snpStringObject;
}



function getAlignmentValue(){


    // #mRNA_alignment
//     const alignmentText = `                                                                                                                                                                                                                                                                                                                                                                                                                                                                            ACTCAAAACTCACAAACCTCAA
//                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | |   ||||||   |||  
// AGACACGTGCTGGGGCGGGCAGGCGAGCGCCTCAGTCTGGTCGCCTGGCGGTGCCTCCGGCCCCAACGCGCCCGGGCCGCCGCGGGCCGCGCGCGCCGATGCCCGGCTGAGTCACTGGCAGGGCAGCGCGCGTGTGGGAAGGGGCGGAGGGAGTGCGGCCGGCGGGCGGGCGGGGCGCTGGGCTCAGCCCGGCCGCAGGTGACCCGGAGGCCCTCGCCGCCCGCGGCGCCCCGAGCGCTTTGTGAGCAGATGCGGAGCCGAGTGGAGGGCGCGAGCCAGATGCGGGGCGACAGCTGACTTGCTGAGAGGAGGCGGGGAGGCGCGGAGCGCGCGTGTGGTCCTTGCGCCGCTGACTTCTCCACTGGTTCCTGGGCACCGAAAGATAAACCTCTCATAATGAAGGCCCCCGCTGTGCTTGCACCTGGCATCCTCGTGCTCCTGTTTACCTTGGTGCAGAGGAGCAATGGGGAGTGTAAAGAGGCACTAGCAAAGTCCGAGATGAATGTGAATATGAAGTATCAGCTTCCCAACTTCACCGCGGAAACACCCATCCAGAATGTCATTCTACATGAGCATCACATTTTCCTTGGTGCCACTAACTACATTTATGTTTTAAATGAGGAAGACCTTCAGAAGGTTGCTGAGTACAAGACTGGGCCTGTGCTGGAACACCCAGATTGTTTCCCATGTCAGGACTGCAGCAGCAAAGCCAATTTATCAGGAGGTGTTTGGAAAGATAACATCAACATGGCTCTAGTTGTCGACACCTACTATGATGATCAACTCATTAGCTGTGGCAGCGTCAACAGAGGGACCTGCCAGCGACATGTCTTTCCCCACAATCATACTGCTGACATACAGTCGGAGGTTCACTGCATATTCTCCCCACAGATAGAAGAGCCCAGCCAGTGTCCTGACTGTGTGGTGAGCGCCCTGGGAGCCAAAGTCCTTTCATCTGTAAAGGACCGGTTCATCAACTTCTTTGTAGGCAATACCATAAATTCTTCTTATTTCCCAGATCATCCATTGCATTCGATATCAGTGAGAAGGCTAAAGGAAACGAAAGATGGTTTTATGTTTTTGACGGACCAGTCCTACATTGATGTTTTACCTGAGTTCAGAGATTCTTACCCCATTAAGTATGTCCATGCCTTTGAAAGCAACAATTTTATTTACTTCTTGACGGTCCAAAGGGAAACTCTAGATGCTCAGACTTTTCACACAAGAATAATCAGGTTCTGTTCCATAAACTCTGGATTGCATTCCTACATGGAAATGCCTCTGGAGTGTATTCTCACAGAAAAGAGAAAAAAGAGATCCACAAAGAAGGAAGTGTTTAATATACTTCAGGCTGCGTATGTCAGCAAGCCTGGGGCCCAGCTTGCTAGACAAATAGGAGCCAGCCTGAATGATGACATTCTTTTCGGGGTGTTCGCACAAAGCAAGCCAGATTCTGCCGAACCAATGGATCGATCTGCCATGTGTGCATTCCCTATCAAATATGTCAACGACTTCTTCAACAAGATCGTCAACAAAAACAATGTGAGATGTCTCCAGCATTTTTACGGACCCAATCATGAGCACTGCTTTAATAGGACACTTCTGAGAAATTCATCAGGCTGTGAAGCGCGCCGTGATGAATATCGAACAGAGTTTACCACAGCTTTGCAGCGCGTTGACTTATTCATGGGTCAATTCAGCGAAGTCCTCTTAACATCTATATCCACCTTCATTAAAGGAGACCTCACCATAGCTAATCTTGGGACATCAGAGGGTCGCTTCATGCAGGTTGTGGTTTCTCGATCAGGACCATCAACCCCTCATGTGAATTTTCTCCTGGACTCCCATCCAGTGTCTCCAGAAGTGATTGTGGAGCATACATTAAACCAAAATGGCTACACACTGGTTATCACTGGGAAGAAGATCACGAAGATCCCATTGAATGGCTTGGGCTGCAGACATTTCCAGTCCTGCAGTCAATGCCTCTCTGCCCCACCCTTTGTTCAGTGTGGCTGGTGCCACGACAAATGTGTGCGATCGGAGGAATGCCTGAGCGGGACATGGACTCAACAGATCTGTCTGCCTGCAATCTACAAGGTTTTCCCAAATAGTGCACCCCTTGAAGGAGGGACAAGGCTGACCATATGTGGCTGGGACTTTGGATTTCGGAGGAATAATAAATTTGATTTAAAGAAAACTAGAGTTCTCCTTGGAAATGAGAGCTGCACCTTGACTTTAAGTGAGAGCACGATGAATACATTGAAATGCACAGTTGGTCCTGCCATGAATAAGCATTTCAATATGTCCATAATTATTTCAAATGGCCACGGGACAACACAATACAGTACATTCTCCTATGTGGATCCTGTAATAACAAGTATTTCGCCGAAATACGGTCCTATGGCTGGTGGCACTTTACTTACTTTAACTGGAAATTACCTAAACAGTGGGAATTCTAGACACATTTCAATTGGTGGAAAAACATGTACTTTAAAAAGTGTGTCAAACAGTATTCTTGAATGTTATACCCCAGCCCAAACCATTTCAACTGAGTTTGCTGTTAAATTGAAAATTGACTTAGCCAACCGAGAGACAAGCATCTTCAGTTACCGTGAAGATCCCATTGTCTATGAAATTCATCCAACCAAATCTTTTATTAGTACTTGGTGGAAAGAACCTCTCAACATTGTCAGTTTTCTATTTTGCTTTGCCAGTGGTGGGAGCACAATAACAGGTGTTGGGAAAAACCTGAATTCAGTTAGTGTCCCGAGAATGGTCATAAATGTGCATGAAGCAGGAAGGAACTTTACAGTGGCATGTCAACATCGCTCTAATTCAGAGATAATCTGTTGTACCACTCCTTCCCTGCAACAGCTGAATCTGCAACTCCCCCTGAAAACCAAAGCCTTTTTCATGTTAGATGGGATCCTTTCCAAATACTTTGATCTCATTTATGTACATAATCCTGTGTTTAAGCCTTTTGAAAAGCCAGTGATGATCTCAATGGGCAATGAAAATGTACTGGAAATTAAGGGAAATGATATTGACCCTGAAGCAGTTAAAGGTGAAGTGTTAAAAGTTGGAAATAAGAGCTGTGAGAATATACACTTACATTCTGAAGCCGTTTTATGCACGGTCCCCAATGACCTGCTGAAATTGAACAGCGAGCTAAATATAGAGTGGAAGCAAGCAATTTCTTCAACCGTCCTTGGAAAAGTAATAGTTCAACCAGATCAGAATTTCACAGGATTGATTGCTGGTGTTGTCTCAATATCAACAGCACTGTTATTACTACTTGGGTTTTTCCTGTGGCTGAAAAAGAGAAAGCAAATTAAAGATCTGGGCAGTGAATTAGTTCGCTACGATGCAAGAGTACACACTCCTCATTTGGATAGGCTTGTAAGTGCCCGAAGTGTAAGCCCAACTACAGAAATGGTTTCAAATGAATCTGTAGACTACCGAGCTACTTTTCCAGAAGATCAGTTTCCTAATTCATCTCAGAACGGTTCATGCCGACAAGTGCAGTATCCTCTGACAGACATGTCCCCCATCCTAACTAGTGGGGACTCTGATATATCCAGTCCATTACTGCAAAATACTGTCCACATTGACCTCAGTGCTCTAAATCCAGAGCTGGTCCAGGCAGTGCAGCATGTAGTGATTGGGCCCAGTAGCCTGATTGTGCATTTCAATGAAGTCATAGGAAGAGGGCATTTTGGTTGTGTATATCATGGGACTTTGTTGGACAATGATGGCAAGAAAATTCACTGTGCTGTGAAATCCTTGAACAGAATCACTGACATAGGAGAAGTTTCCCAATTTCTGACCGAGGGAATCATCATGAAAGATTTTAGTCATCCCAATGTCCTCTCGCTCCTGGGAATCTGCCTGCGAAGTGAAGGGTCTCCGCTGGTGGTCCTACCATACATGAAACATGGAGATCTTCGAAATTTCATTCGAAATGAGACTCATAATCCAACTGTAAAAGATCTTATTGGCTTTGGTCTTCAAGTAGCCAAAGGCATGAAATATCTTGCAAGCAAAAAGTTTGTCCACAGAGACTTGGCTGCAAGAAACTGTATGCTGGATGAAAAATTCACAGTCAAGGTTGCTGATTTTGGTCTTGCCAGAGACATGTATGATAAAGAATACTATAGTGTACACAACAAAACAGGTGCAAAGCTGCCAGTGAAGTGGATGGCTTTGGAAAGTCTGCAAACTCAAAAGTTTACCACCAAGTCAGATGTGTGGTCCTTTGGCGTGCTCCTCTGGGAGCTGATGACAAGAGGAGCCCCACCTTATCCTGACGTAAACACCTTTGATATAACTGTTTACTTGTTGCAAGGGAGAAGACTCCTACAACCCGAATACTGCCCAGACCCCTTATATGAAGTAATGCTAAAATGCTGGCACCCTAAAGCCGAAATGCGCCCATCCTTTTCTGAACTGGTGTCCCGGATATCAGCGATCTTCTCTACTTTCATTGGGGAGCACTATGTCCATGTGAACGCTACTTATGTGAACGTAAAATGTGTCGCTCCGTATCCTTCTCTGTTGTCATCAGAAGATAACGCTGATGATGAGGTGGACACACGACCAGCCTCCTTCTGGGAGACATCATAGTGCTAGTACTATGTCAAAGCAACAGTCCACACTTTGTCCAATGGTTTTTTCACTGCCTGACCTTTAAAAGGCCATCGATATTCTTTGCTCTTGCCAAAATTGCACTATTATAGGACTTGTATTGTTATTTAAATTACTGGATTCTAAGGAATTTCTTATCTGACAGAGCATCAGAACCAGAGGCTTGGTCCCACAGGCCACGGACCAATGGCCTGCAGCCGTGACAACACTCCTGTCATATTGGAGTCCAAAACTTGAATTCTGGGTTGAATTTTTTAAAAATCAGGTACCACTTGATTTCATATGGGAAATTGAAGCAGGAAATATTGAGGGCTTCTTGATCACAGAAAACTCAGAAGAGATAGTAATGCTCAGGACAGGAGCGGCAGCCCCAGAACAGGCCACTCATTTAGAATTCTAGTGTTTCAAAACACTTTTGTGTGTTGTATGGTCAATAACATTTTTCATTACTGATGGTGTCATTCACCCATTAGGTAAACATTCCCTTTTAAATGTTTGTTTGTTTTTTGAGACAGGATCTCACTCTGTTGCCAGGGCTGTAGTGCAGTGGTGTGATCATAGCTCACTGCAACCTCCACCTCCCAGGCTCAAGCCTCCCGAATAGCTGGGACTACAGGCGCACACCACCATCCCCGGCTAATTTTTGTATTTTTTGTAGAGACGGGGTTTTGCCATGTTGCCAAGGCTGGTTTCAAACTCCTGGACTCAAGAAATCCACCCACCTCAGCCTCCCAAAGTGCTAGGATTACAGGCATGAGCCACTGCGCCCAGCCCTTATAAATTTTTGTATAGACATTCCTTTGGTTGGAAGAATATTTATAGGCAATACAGTCAAAGTTTCAAAATAGCATCACACAAAACATGTTTATAAATGAACAGGATGTAATGTACATAGATGACATTAAGAAAATTTGTATGAAATAATTTAGTCATCATGAAATATTTAGTTGTCATATAAAAACCCACTGTTTGAGAATGATGCTACTCTGATCTAATGAATGTGAACATGTAGATGTTTTGTGTGTATTTTTTTAAATGAAAACTCAAAATAAGACAAGTAATTTGTTGATAAATATTTTTAAAGATAACTCAGCATGTTTGTAAAGCAGGATACATTTTACTAAAAGGTTCATTGGTTCCAATCACAGCTCATAGGTAGAGCAAAGAAAGGGTGGATGGATTGAAAAGATTAGCCTCTGTCTCGGTGGCAGGTTCCCACCTCGCAAGCAATTGGAAACAAAACTTTTGGGGAGTTTTATTTTGCATTAGGGTGTGTTTTATGTTAAGCAAAACATACTTTAGAAACAAATGAAAAAGGCAATTGAAAATCCCAGCTATTTCACCTAGATGGAATAGCCACCCTGAGCAGAACTTTGTGATGCTTCATTCTGTGGAATTTTGTGCTTGCTACTGTATAGTGCATGTGGTGTAGGTTACTCTAACTGGTTTTGTCGACGTAAACATTTAAAGTGTTATATTTTTTATAAAAATGTTTATTTTTAATGATATGAGAAAAATTTTGTTAGGCCACAAAAACACTGCACTGTGAACATTTTAGAAAAGGTATGTCAGACTGGGATTAATGACAGCATGATTTTCAATGACTGTAAATTGCGATAAGGAAATGTACTGATTGCCAATACACCCCACCCTCATTACATCATCAGGACTTGAAGCCAAGGGTTAACCCAGCAAGCTACAAAGAGGGTGTGTCACACTGAAACTCAATAGTTGAGTTTGGCTGTTGTTGCAGGAAAATGATTATAACTAAAAGCTCTCTGATAGTGCAGAGACTTACCAGAAGACACAAGGAATTGTACTGAAGAGCTATTACAATCCAAATATTGCCGTTTCATAAATGTAATAAGTAATACTAATTCACAGAGTATTGTAAATGGTGGATGACAAAAGAAAATCTGCTCTGTGGAAAGAAAGAACTGTCTCTACCAGGGTCAAGAGCATGAACGCATCAATAGAAAGAACTCGGGGAAACATCCCATCAACAGGACTACACACTTGTATATACATTCTTGAGAACACTGCAATGTGAAAATCACGTTTGCTATTTATAAACTTGTCCTTAGATTAATGTGTCTGGACAGATTGTGGGAGTAAGTGATTCTTCTAAGAATTAGATACTTGTCACTGCCTATACCTGCAGCTGAACTGAATGGTACTTCGTATGTTAATAGTTGTTCTGATAAATCATGCAATTAAAGTAAAGTGATGCAA
    
// `;
    const alignmentText = $('#mRNA_alignment').val();

    const lines = alignmentText.split('\n');
    const mRNA_line = lines[0];
    const matchLine = lines[1];
    // count leading white space in first line
    const mRNA_leadingWhiteSpace = mRNA_line.match(/^\s*/)[0].length;
    const match_leadingWhiteSpace = matchLine.match(/^\s*/)[0].length;

    console.log(mRNA_leadingWhiteSpace, match_leadingWhiteSpace)

    //return alignmentText;

    return {
        alignment: alignmentText,
        mRNA_leadingWhiteSpace: mRNA_leadingWhiteSpace,
        match_leadingWhiteSpace: match_leadingWhiteSpace,
        lines: lines,
        mRNA_line: mRNA_line,
        seq_length: mRNA_line.replace(/\s/g, '').length,
        matchLine: matchLine,
    }
}


getAlignmentValue();

function copyFasta(){
    // Get the text field
    if(currentTranscript.fasta=='' || currentTranscript.fasta==undefined){
        return;
    }

    var copyText = document.getElementById("copyInput");
    copyText.value = currentTranscript.fasta;
    

    // Select the text field
    copyText.select();
    copyText.setSelectionRange(0, 99999); // For mobile devices

    // Copy the text inside the text field
    navigator.clipboard.writeText(copyText.value);
    alert('Copied!');
    
}
function getCurrentSequenceInFasta(){
    
}

</script>
</html>